Add GlusterFS 10.4 release notes (#4101)
gluster/glusterfs: doc/release-notes/10.4.md

# Release notes for Gluster 10.4

This is a bugfix release. The release notes for [10.0](10.0.md), [10.1](10.1.md), [10.2](10.2.md) and [10.3](10.3.md) contain a listing of all the new improvements and bugs fixed in the GlusterFS 10 stable release.

**NOTE:**
- Next minor release tentative date: Week of 25th Sep, 2023
- Users are highly encouraged to upgrade to newer releases of GlusterFS.

## Important fixes in this release
- Fix fuse concurrency problems
- Fix memory corruption in debug builds
- Fix recovery issue with posix locks upon reconnection of a disconnected brick

## Builds are available at

[https://download.gluster.org/pub/gluster/glusterfs/10/10.4/](https://download.gluster.org/pub/gluster/glusterfs/10/10.4/)

## Issues addressed in this release

> - [#1000](https://github.com/gluster/glusterfs/issues/1000) configure: Force 'char' type to be signed inorder to eliminate anomalies

Reviewer: ```suggestion
- [#1000](https://github.com/gluster/glusterfs/issues/1000) configure: Force 'char' type to be signed in order to eliminate anomalies
```

- [#2752](https://github.com/gluster/glusterfs/issues/2752) posix: Fix directory gfid handle if a rename fails
- [#3345](https://github.com/gluster/glusterfs/issues/3345) Fix inconsistencies in big-endian architectures of hashfn

> - [#3346](https://github.com/gluster/glusterfs/issues/3346) Fix stack overflow when processing glx_dir\(p\) list structures in xdr

Reviewer: ```suggestion
- [#3346](https://github.com/gluster/glusterfs/issues/3346) Fix stack overflow when processing glx_dir(p) list structures in xdr
```

> - [#3659](https://github.com/gluster/glusterfs/issues/3659) Fix fuse concurrency problems

Reviewer: This is actually the same as "Fix memory corruption in debug builds" (line 25)

> - [#3855](https://github.com/gluster/glusterfs/issues/3855) Fix memory corruption in debug builds

Reviewer: The patch for this issue included several changes from devel to be able to easily backport the principal patch. However the other patches are mostly cleanups, so probably this entry is not required.

> - [#3882](https://github.com/gluster/glusterfs/issues/3882) Fix deadlock in  gf_print_trace of sys_log

Reviewer: ```suggestion
- [#3882](https://github.com/gluster/glusterfs/issues/3882) Fix deadlock in gf_print_trace of sys_log
```

> - [#3901](https://github.com/gluster/glusterfs/issues/3901) Fix segmentaion fault in  io-stats xlator

Reviewer: ```suggestion
- [#3901](https://github.com/gluster/glusterfs/issues/3901) Fix segmentaion fault in io-stats xlator
```

- [#3954](https://github.com/gluster/glusterfs/issues/3954) Fix stack-buffer-overflow according to AddressSanitizer
- [#4020](https://github.com/gluster/glusterfs/issues/4020) Improve regression test suite
- [#4029](https://github.com/gluster/glusterfs/issues/4029) Process stuck listing snapshots from NFS
- [#4031](https://github.com/gluster/glusterfs/issues/4031) Fix Input/Output error when using linux-aio on big-endean architectures
- [#4042](https://github.com/gluster/glusterfs/issues/4042) Fix recovery issue with posix locks upon reconnection of a disconnected brick

> - [#4071](https://github.com/gluster/glusterfs/issues/4071) Make timestamps stable

Reviewer: ```suggestion
- [#4071](https://github.com/gluster/glusterfs/issues/4071) Make timestamps of .snap directory stable
```

- [#3894](https://github.com/gluster/glusterfs/pull/3894) Use opendir for directories in glfs_open and glfs_h_open of api

> - [#4029](https://github.com/gluster/glusterfs/issues/4029) Process stuck listing snapshots from NFS

Reviewer: This one is duplicated (line 30)
